keps/sig-cloud-provider/20190704-instance-host-labelling.md

```diff
@@ -0,0 +1,174 @@
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Many kubernetes clusters are deployed on virtual machines (also referred to as "instances"),
+and information about the physical host for those instances (also referred to as "instance host")
+is not consistently available.
+Instances which share a host can fail simultaneously if there is a problem with the host.
+
+We propose to add a well-known label for nodes to be filled with the instance host ID.
+This label can then be used by controllers (namely, the scheduler) to implement features such as inter-pod affinity and anti-affinity and even pod spreading.
+
+## Motivation
+
+Today, well-known labels for zone and region can be used as topology keys for inter-pod affinity and anti-affinity.
+There is also ongoing work to implement even pod spreading, for which these keys can also be used.
+
+We want to provide more built-in options to improve reliability of workloads running in a kubernetes cluster.
+An instance host can fail and thus disrupting all of its pods, so it should be identified as a topology or failure domain as well.
+
+In addition to reliability, having a well-defined label for the instance host increases the portability of pod specs for customers that have on-prem and cloud deployments.
+
+### Goals
+
+- Introduce a well known label to identify the instance host of a node.
+- Define a default value for the label so that components maintain their current behavior if they use the label.
+
+### Non-Goals
+
+- Add a cloud provider API to get instance host ID.
+- Populate the instance host label with the instance host ID.
+- Controller behavior (e.g. deployment, statefulset) that considers instance host label.
+- Any decision-making that is based on the physical host label.
+
+## Proposal
+
+Add a well-known label for instance host: `topology.kubernetes.io/instance-host` and make its default value empty.
+
+The label will be named `failure-domain.kubernetes.io/instance-host` if KEP [Promoting cloud provider labels to GA](https://github.com/kubernetes/enhancements/pull/839) doesn’t land.
+
+
+### User Stories
+
+#### Story 1
+
+As an application developer, I want pods to be scheduled in the same instance host so that networking between them is faster.
```

Comment:
Would be great to highlight in which scenarios instance-host is useful over just kubernetes node topology for this case.

Reply:
While I would call this (depending on faster networking by co-scheduling) an anti-pattern I don't think we should hide information or stop people from making their own decisions.

Under no circumstances can we proscribe the semantic implications of a platform when co-scheduling.  It might be faster, it might be the same, or it might be slower.  We can not say anything about that, in the abstract.  A given implementation may choose to document this or not.

Reply:
Story 2 was actually the driver for this KEP #75274. However, even pod spreading is a feature in development, whereas pod (anti)affinity already exists. We can remove this Story if it's something we don't want to promote.

Reply:
This is definitely a story that I don't think we want to promote given the history behind pods (co-scheduling) and existing affinity rules.